tests/Functional/TestCase.php

```diff
@@ -82,16 +86,114 @@ protected static function runFrontScript(array $args, string $cwd = __DIR__): Pr
+        }
+    }
+
+    protected static function createFile(string $baseDir, string $filename): void
+    {
+        $filename = "{$baseDir}/{$filename}";
+        $dirname = dirname($filename);
+        if (!file_exists($dirname)) {
+            self::assertTrue(mkdir($dirname, 0777, true), 'Created directory.');
+        }
+        self::assertTrue(touch($filename), 'Created file.');
+        self::assertNotFalse(realpath($filename), 'Got absolute path.');
+    }
+
+    protected static function changeFile($dir, $filename): void
+    {
+        file_put_contents(
+            DirectoryUtil::ensureTrailingSlash($dir) . $filename,
+            self::CHANGED_CONTENT
+        );
+    }
+
+    protected static function deleteFile($dir, $filename): void
+    {
+        unlink(DirectoryUtil::ensureTrailingSlash($dir) . $filename);
+    }
+
+    protected static function fixSeparators(string $path)
+    {
+        return str_replace(['/', '\\'], DIRECTORY_SEPARATOR, $path);
+    }
+
+    protected static function fixSeparatorsMultiple(&...$paths): void
+    {
+        foreach ($paths as &$path) {
+            $path = self::fixSeparators($path);
+        }
+    }
+
+    protected static function assertDirectoryListing(string $dir, array $expected, string $ignoreDir = ''): void
+    {
+        $ignoreDir = DirectoryUtil::stripAncestor($ignoreDir, $dir);
+
+        $expected = array_map(static function ($path) {
+            return self::fixSeparators($path);
+        }, $expected);
+
+        $expected = array_filter($expected);
+        asort($expected);
+        $expected = array_values($expected);
+
+        $actual = self::getFlatDirectoryListing($dir);
+        $actual = array_map(static function ($path) use ($ignoreDir) {
+            $ignoreDir = DirectoryUtil::ensureTrailingSlash($ignoreDir);
+            if (strpos($path, $ignoreDir) === 0) {
+                return false;
             }
-            self::assertTrue(touch($filename), 'Created file.');
-            self::assertNotFalse(realpath($filename), 'Got absolute path.');
+            return self::fixSeparators($path);
+        }, $actual);
+
+        $actual = array_filter($actual);
+        asort($actual);
+        $actual = array_values($actual);
+
+        self::assertEquals($expected, $actual, "Directory {$dir} contains the expected files.");
+    }
+
+    protected static function getFlatDirectoryListing(string $dir): array
+    {
+        $dir = DirectoryUtil::stripTrailingSlash($dir);
+
+        $iterator = new RecursiveIteratorIterator(
+            new \RecursiveDirectoryIterator($dir)
+        );
+
+        $listing = [];
+        /** @var \SplFileInfo $splFileInfo */
+        foreach ($iterator as $splFileInfo) {
+            if (in_array($splFileInfo->getFilename(), ['.', '..'])) {
+                continue;
+            }
+            $pathname = $splFileInfo->getPathname();
+            $listing[] = substr($pathname, strlen($dir) + 1);
```

Comment:
It would be helpful to have a comment here with an example of the output this operation will produce. It's hard to visualize it from a reading of the code.

Reply:
Sure. I've added a docblock with such an example.